File: Data Structures & Algorithms/binary-search/submission-0.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        
        if len(nums) == 0: return -1

        if len(nums) == 1: 
            if nums[0] == target:
                return 0
            return -1

        l, r = 0, len(nums) - 1

        while l <= r:
            
            # get mid and check target
            mid_i = (r+l)//2

            if nums[mid_i] == target: return mid_i

            if nums[mid_i] < target:
                l = mid_i+1
            else:
                r = mid_i-1

        return -1

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-0.cs
public class Solution {
    public int Search(int[] nums, int target) {
        
        int l = 0;
        int r = nums.Length-1;
        int mid;
        while(l<=r){

            mid = l+(r-l)/2;

            // base case
            if (nums[mid] == target) return mid;

            // is left sorted locally
            if(nums[l]<=nums[mid]){
                // check for value
                if(nums[l]<=target && target<nums[mid]){
                    r = mid-1;
                }
                // if not here then checkother half
                else{
                    l = mid+1;
                }
            }
            // right unsorted
            else{
                // check for value
                if(nums[mid]<=target && target<=nums[r]){
                    l = mid+1;
                }
                // if not here then checkother half
                else{
                    r = mid-1;
                }
            }
        }

        return -1;
    }
}
